contracts/crashpunks-v2-friedger.clar

```diff
@@ -0,0 +1,345 @@
+        (ok (var-set administrator new-administrator))
+    )
+)
+
+(define-public (set-collection-royalties (new-mint-addresses (list 4 principal)) (new-mint-shares (list 4 uint)) (new-royalty-addresses (list 10 principal)) (new-royalty-shares (list 10 uint)))
+    (begin
+        (asserts! (is-eq (var-get administrator) contract-caller) ERR-NOT-ADMINISTRATOR)
+        (var-set collection-mint-addresses new-mint-addresses)
+        (var-set collection-mint-shares new-mint-shares)
+        (var-set collection-royalty-addresses new-royalty-addresses)
+        (var-set collection-royalty-shares new-royalty-shares)
+        (ok true)
+    )
+)
+
+(define-public (set-base-uri (new-base-uri (string-ascii 80)))
+    (begin
+        (asserts! (is-eq contract-caller (var-get administrator)) ERR-NOT-ADMINISTRATOR)
+        (asserts! (not (var-get metadata-frozen)) ERR-METADATA-FROZEN)
+        (var-set base-uri new-base-uri)
+        (ok true))
+)
+
+(define-public (freeze-metadata)
+    (begin
+        (asserts! (is-eq contract-caller (var-get administrator)) ERR-NOT-ADMINISTRATOR)
+        (var-set metadata-frozen true)
+        (ok true)
+    )
+)
+
+;; read only methods
+(define-read-only (get-token-market-by-index (nftIndex uint))
+    (map-get? nft-market nftIndex)
+)
+
+(define-read-only (get-mint-pass-balance (account principal))
+    (default-to u0
+        (map-get? mint-pass account)
+    )
+)
+
+
+;; private methods
+(define-private (is-owned-or-approved (nftIndex uint) (operator principal) (owner principal))
+    (default-to 
+        (default-to
+            (is-eq owner operator)
+            (map-get? approvals-all {owner: owner, operator: operator})
+        )
+        (map-get? approvals {owner: owner, operator: operator, nft-index: nftIndex})
+    )
+)
+
+(define-private (paymint-split (mintPrice uint) (payer principal)) 
+    (let (
+            (split u0)
+            (mintAddresses (var-get collection-mint-addresses))
+            (mintShares (var-get collection-mint-shares))
+        )
+        (+ split (unwrap! (pay-royalty payer mintPrice (unwrap! (element-at mintAddresses u0) ERR-PAYMENT-ADDRESS) (unwrap! (element-at mintShares u0) ERR-PAYMENT-ADDRESS)) ERR-PAYMENT-ADDRESS))
+        (+ split (unwrap! (pay-royalty payer mintPrice (unwrap! (element-at mintAddresses u1) ERR-PAYMENT-ADDRESS) (unwrap! (element-at mintShares u1) ERR-PAYMENT-ADDRESS)) ERR-PAYMENT-ADDRESS))
+        (+ split (unwrap! (pay-royalty payer mintPrice (unwrap! (element-at mintAddresses u2) ERR-PAYMENT-ADDRESS) (unwrap! (element-at mintShares u2) ERR-PAYMENT-ADDRESS)) ERR-PAYMENT-ADDRESS))
+        (+ split (unwrap! (pay-royalty payer mintPrice (unwrap! (element-at mintAddresses u3) ERR-PAYMENT-ADDRESS) (unwrap! (element-at mintShares u3) ERR-PAYMENT-ADDRESS)) ERR-PAYMENT-ADDRESS))
+        (ok split)
+    )
+)
+
+;; TODO: update this to use a list of {address, share}, and fold 
```

Comment:
Is fold more efficient?

Reply:
i was intending to just use fold so that the length of the list could be flexible, vs always having to pass in a list of 10